x-pack/metricbeat/module/aws/awshealth/awshealth.go

```diff
@@ -0,0 +1,431 @@
+	if v != nil {
+		return *v
+	}
+	return defaultValue
+}
+
+func createAffectedEntityDetails(affectedEntities []types.AffectedEntity) []AffectedEntityDetails {
+	aed := make([]AffectedEntityDetails, len(affectedEntities))
+	for i := range affectedEntities {
+		aed[i] = AffectedEntityDetails{
+			AwsAccountId:    getValueOrDefault(affectedEntities[i].AwsAccountId, ""),
+			EntityUrl:       getValueOrDefault(affectedEntities[i].EntityUrl, ""),
+			EntityValue:     getValueOrDefault(affectedEntities[i].EntityValue, ""),
+			LastUpdatedTime: getValueOrDefault(affectedEntities[i].LastUpdatedTime, time.Time{}),
+			StatusCode:      getValueOrDefault((*string)(&affectedEntities[i].StatusCode), ""),
+			EntityArn:       getValueOrDefault(affectedEntities[i].EntityArn, ""),
+		}
+	}
+	return aed
+}
+
+// generateEventID hashes the provided eventID and returns the first 20 characters.
+// This is used to generate a unique but consistent event ID prefix.
+func generateEventID(eventID string) string {
+	h := sha256.New()
+	h.Write([]byte(eventID))
+	prefix := hex.EncodeToString(h.Sum(nil))
+	return prefix[:20]
+}
+
+// getEventsSummary retrieves a summary of AWS Health events that meet the specified
+// filter criteria. It uses the DescribeEvents API to get a list of events. For each
+// event, it calls getDescribeEventDetails with the EventARN to get details like details
+// of affected entities by calling DescribeAffectedEntities API. The DescribeAffectedEntities
+// is called to fetch the description of the event.
+// The function returns a slice of mb.Event structs containing the summarized event info.
+func (m *MetricSet) getDescribeEventDetails(ctx context.Context, awsHealth *health.Client, event types.Event, ch chan<- HealthDetails) error {
+	hd := HealthDetails{event: event, affectedEntities: []types.AffectedEntity{}}
+	eventDetails, err := awsHealth.DescribeEventDetails(ctx, &health.DescribeEventDetailsInput{
+		EventArns: []string{*event.Arn},
+		Locale:    &locale,
+	})
+	if err != nil {
+		if errors.Is(err, context.Canceled) {
+			m.Logger().Debug("Context cancelled. Exiting gracefully.")
+			return nil
+		}
+		err = fmt.Errorf("[AWS Health] DescribeEventDetails failed with : %w", err)
+		return err
+	}
+
+	if len(eventDetails.SuccessfulSet) > 0 {
+		hd.eventDescription = *(eventDetails.SuccessfulSet[0].EventDescription.LatestDescription)
+	} else {
+		hd.eventDescription = "Unable to find event description details."
+	}
+
+	var (
+		affEntityTokString string
+		nextToken          *string
+		pending            int32
+		resolved           int32
+		others             int32
+	)
+
+	for {
+		// When invoking the DescribeAffectedEntities for the first time, there must not exist any NextToken.
+		// DescribeAffectedEntities API call will return the next token if there are more records left for querying
+		// If there exist no further records to fetch, next toke will be empty.
+		if affEntityTokString == "" {
```

Comment:
there is a tonne of duplication here. (and again, is there a reason not to use the built in paginator?)